### PR TITLE
[Lang] Support configure sparse solver ordering

### PR DESCRIPTION
**Pull request by FantasyVR**
Related issue = https://github.com/taichi-dev/taichi/issues/2906

### CODE REVIEW

File: taichi/program/sparse_solver.cpp [hunk 1]
```diff
@@ -32,17 +32,37 @@ bool EigenSparseSolver<EigenSolver>::info() {
   return solver_.info() == Eigen::Success;
 }
 
-std::unique_ptr<SparseSolver> get_sparse_solver(
-    const std::string &solver_type) {
-  if (solver_type == "LU") {
+#include <map>
```
Comment (k-ye) [inline taichi/program/sparse_solver.cpp:35]: move to top

File: taichi/program/sparse_solver.cpp [hunk 2]
```diff
@@ -32,17 +32,37 @@ bool EigenSparseSolver<EigenSolver>::info() {
...
+  {                                                                           \
+    {#type, #order}, []() -> std::unique_ptr<SparseSolver> {                  \
+      using T =                                                               \
+          Eigen::Simplicial##type<Eigen::SparseMatrix<float32>, Eigen::Lower, \
+                                  Eigen::order##Ordering<int>>;               \
+      return std::make_unique<EigenSparseSolver<T>>();                        \
+    }                                                                         \
+  }
+
+static std::map<key_type, func_type> solver_factory = {
```
Comment (k-ye) [inline taichi/program/sparse_solver.cpp:49]: nit: move this into `get_sparse_solver` and make it const? use unordered_map

**Review by k-ye — COMMENTED** [review]
Cool! A few nits

File: taichi/program/sparse_solver.cpp (hunk 2 above)
Comment (FantasyVR) [reply taichi/program/sparse_solver.cpp:49]: OK, I could move `solver_factory` into  `get_sparse_solver`.

As far as I know, `unordered_map` couldn't hash `std::pair`.

File: taichi/program/sparse_solver.cpp (hunk 1 above)
Comment (FantasyVR) [reply taichi/program/sparse_solver.cpp:35]: Thanks

File: taichi/program/sparse_solver.cpp (hunk 2 above)
Comment (k-ye) [reply taichi/program/sparse_solver.cpp:49]: Yeah, you can provide a customized hasher object for that specific type of pair. See the third template type of https://en.cppreference.com/w/cpp/container/unordered_map :-)

File: taichi/program/sparse_solver.cpp [hunk 3]
```diff
@@ -32,17 +33,44 @@ bool EigenSparseSolver<EigenSolver>::info() {
...
+  {                                                                           \
+    {#type, #order}, []() -> std::unique_ptr<SparseSolver> {                  \
+      using T =                                                               \
+          Eigen::Simplicial##type<Eigen::SparseMatrix<float32>, Eigen::Lower, \
+                                  Eigen::order##Ordering<int>>;               \
+      return std::make_unique<EigenSparseSolver<T>>();                        \
+    }                                                                         \
+  }
+
+struct pair_hash {
```
Comment (k-ye) [inline taichi/program/sparse_solver.cpp:46]: Final nit: place this at the top and guard it with an anonymous namespace (e.g. `namespace { ... } // namespace`)

**Review by k-ye — APPROVED** [review]
LGTM!

**Pull request merged**